smithy-trait-codegen/codegen-core/src/main/java/software/amazon/smithy/traitcodegen/generators/traits/DefaultTraitGeneratorProvider.java

```diff
@@ -0,0 +1,118 @@
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.ShortShape;
+import software.amazon.smithy.model.shapes.StringShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.traitcodegen.TraitGeneratorProvider;
+
+public final class DefaultTraitGeneratorProvider extends ShapeVisitor.Default<TraitGenerator>
+        implements TraitGeneratorProvider {
+    @Override
+    public TraitGenerator getGenerator(Shape shape) {
+        return shape.accept(this);
+    }
+
+    @Override
+    protected TraitGenerator getDefault(Shape shape) {
+        throw new UnsupportedOperationException("Trait code generation does not support shapes of type: "
+                + shape.getType());
+    }
+
+    @Override
+    public TraitGenerator booleanShape(BooleanShape shape) {
+        throw new UnsupportedOperationException("Boolean shapes not supported for trait code generation. "
+                + "Consider using an Annotation trait instead");
+    }
+
+    @Override
+    public TraitGenerator intEnumShape(IntEnumShape shape) {
+        return new IntEnumTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator stringShape(StringShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator enumShape(EnumShape shape) {
+        return new EnumTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator listShape(ListShape shape) {
+        return new CollectionTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator byteShape(ByteShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator shortShape(ShortShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator integerShape(IntegerShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator longShape(LongShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator floatShape(FloatShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator documentShape(DocumentShape shape) {
+        return new DocumentTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator doubleShape(DoubleShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator bigDecimalShape(BigDecimalShape shape) {
+        return new ValueTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator mapShape(MapShape shape) {
+        return new CollectionTraitGenerator();
+    }
+
+    @Override
+    public TraitGenerator structureShape(StructureShape shape) {
+        if (shape.getAllMembers().isEmpty()) {
+            return new AnnotationTraitGenerator();
```

Comment:
This is potentially risky since adding members to the trait would make it a breaking change on the java side. Definitely log this, maybe add a setting to disable this behavior.

Reply:
That is a good point. I think we might want to always have annotation traits be just normal structure traits for that reason. With the boilerplate being code generated inheriting the AnnotationTrait base class becomes less necessary/helpful. Plus that would remove a bunch of conditional branches in the code gen. I will test to make sure the structure trait handles an empty struct well and if so switch over.

Reply:
I removed all usage of the Annotation trait generator in commit [281f487](https://github.com/smithy-lang/smithy/pull/2074/commits/281f4876c2dfd05efe0ccd55ca6b87c0618165e1). It also had a nice effect of simplifying some of the code.